### PR TITLE
feat(toasts): Toasts overhaul

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

During work on #8681 it came up that our toasts are all over the place. This overhauls them based on [draft design](https://www.figma.com/file/Y9G24U4r04nEjIDGIEGuKI/PostHog-Design-System-One?node-id=1813%3A518).

| Before | After |
| --- | --- |
| Screen Shot 2022-02-22 at 10 31 10 | Screen Shot 2022-02-22 at 10 36 01 |

### CODE REVIEW

**Comment by liyiy** [comment]
lemon toast? : )

File: frontend/src/scenes/App.tsx [hunk 1]
```diff
@@ -111,7 +117,15 @@ function AppScene(): JSX.Element | null {
         (activeScene ? loadedScenes[activeScene]?.component : null) ||
         (() => (showingDelayedSpinner ? <SceneLoading /> : null))
 
-    const toastContainer = <ToastContainer autoClose={8000} transition={Slide} position="bottom-right" />
+    const toastContainer = (
+        <ToastContainer
+            autoClose={8000}
```
Comment (clarkus) [inline frontend/src/scenes/App.tsx:122]: These are critical attributes for this kind of component. The amount of time the toast is show, the animation speed and duration as it moves in and out of view, etc. Typically I cap these durations at 7 seconds or less. The slide transition might make these feel slower in some ways - were there any other transition or animation types you tried that felt just as effective but possibly snappier of faster?

**Review by clarkus — COMMENTED** [review]
Awesome start to the toast changes. How do toasts stack up when multiple exist concurrently? Depending on user actions and such, we might show multiple of these. Is there an upper bound to how many we show concurrently? Do you think this is a problem worth exploring with this change, or is the scope focused mostly on aligning things visually?

**Comment by Twixes** [comment]
Alright, I've updated ALL the toasts for the new convention ([updated Figma designs](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7854%3A63501)):
- there always is an icon (usually just the default success/info/error one, as I'm temporarily tired of looking at hundreds of toasts)
- the message is always only a single sentence
- there can be a single action button – for almost all error buttons it's the "Get help" one, otherwise _some_ success/info ones have "View dashboard" or something like that
- reduced presence from 8 to 6 seconds
- they can't be dragged sideways anymore because usually that behavior just got in the way – should be a positive change as the close button is better exposed now

Some examples:
Screen Shot 2022-03-04 at 16 04 26
Screen Shot 2022-03-04 at 16 04 36
Screen Shot 2022-03-04 at 16 04 55

**Review by clarkus — APPROVED** [review]
Nailed it 🏆

File: cypress/integration/cohorts.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ describe('Cohorts', () => {
 
         // save
         cy.get('[data-attr="save-cohort"]').click()
-        cy.get('[data-attr=success-toast]').should('exist')
+        cy.get('.Toastify__toast').contains('Cohort saved').should('exist')
```
Comment (mariusandra) [inline cypress/integration/cohorts.js:30]: Should we keep the `data-attr` on the toasts? This could be used for instrumentation somewhere... and is programmatically better than relying on styling hints (css classes) when pinpointing content... 🤔

File: frontend/src/lib/components/lemonToast.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            )}
+        </div>
+    )
+}
+
+function ensureToastId(toastOptions: ToastOptions): void {
+    if (!toastOptions.toastId) {
+        toastOptions.toastId = `lemon-${Math.round(Math.random() * 10000000)}`
+    }
+}
```
Comment (mariusandra) [inline frontend/src/lib/components/lemonToast.tsx:52]: Nit: I'd add an explicit comment here why it's fine to mutate the object (`...toastOptions`), as opposed to all the other places where we default to disallowing this kind of behaviour. Alternatively, rewriting this in a functional way (returns a new object) won't really cost us much performance.

**Review by mariusandra — APPROVED** [review]
Looks good, just two nits/comments inline.

File: frontend/src/scenes/feature-flags/FeatureFlags.tsx [hunk 4]
```diff
@@ -122,7 +122,7 @@ export function FeatureFlags(): JSX.Element {
                                         onClick={() => {
                                             deleteWithUndo({
                                                 endpoint: `projects/${currentTeamId}/feature_flags`,
-                                                object: { name: featureFlag.name, id: featureFlag.id },
+                                                object: { name: featureFlag.key, id: featureFlag.id },
```
Comment (pauldambra) [inline frontend/src/scenes/feature-flags/FeatureFlags.tsx:125]: this seems like an unrelated change...

**Review by pauldambra — APPROVED** [review]

**Comment by pauldambra** [comment]
also, might be a red herring caused by switching branches but I see errors inside the lemon toast code

Screenshot 2022-03-07 at 09 19 02

File: frontend/src/scenes/feature-flags/FeatureFlags.tsx (hunk 4 above)
Comment (Twixes) [reply frontend/src/scenes/feature-flags/FeatureFlags.tsx:125]: It's a definite fix, but I think it fits into the "overhaul" goal. So for feature flags, the `name` field means the description, while the _actual_ "name" of the FF is its `key`. 🤯 Noticed this when testing out the new style

File: frontend/src/lib/components/lemonToast.tsx (hunk 3 above)
Comment (Twixes) [reply frontend/src/lib/components/lemonToast.tsx:52]: Fair, converted it to the functional way

**Comment by Twixes** [comment]
I think you may just need to rerun `yarn` @pauldambra because this PR upgrades `react-toastify`.

**Comment by Twixes** [comment]
I'm being blocked here by these Jest test failures – somehow connected to Kea, even though the `lemonToast` interface doesn't use Kea 🤔 

Screen Shot 2022-03-07 at 15 43 07

Kea Support Hero @mariusandra

**Comment by mariusandra** [comment]
My gut feeling says there's probably a roundabout import. Something like:

```ts
// a.ts
import { b } from 'b'
export const a = 'woot'

// b.ts
import { a } from 'a'
export const b = 'hoot'
```

I'd be happy to investigate later, but knee deep in some service worker logic now... :)

**Comment by mariusandra** [comment]
Seems like `deinitionPopupLogic` depends on `preflightLogic`, which depends on `teamLogic`, which depends on `organizationLogic`, which depends on `userLogic`, which depends on `preflightLogic`.

The new connection is preflightLogic importing userLogic here: https://github.com/PostHog/posthog/pull/8710/files#diff-69c13b4456cc0f6e8b568d27035e21c315e96ea764dd70cf26a98807f799bf50R10 ... but removing it made no difference. 🤔
